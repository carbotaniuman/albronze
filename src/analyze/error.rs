use thiserror::Error;

use crate::InternedStr;

use crate::data::{ComparisonToken, StorageClass};
use crate::hir::*;
use crate::parse::ast::*;

impl<S: Into<String>> From<S> for SemanticError {
    fn from(err: S) -> Self {
        SemanticError::Generic(err.into())
    }
}

#[derive(Clone, Debug, Error, PartialEq)]
pub enum Error {
    #[error("invalid program: {0}")]
    Semantic(#[from] SemanticError),

    #[error("invalid syntax: {0}")]
    Syntax(#[from] crate::parse::error::SyntaxError),

    #[error("invalid macro: {0}")]
    PreProcessor(#[from] crate::preprocess::error::CppError),
}

/// Semantic errors are non-exhaustive and may have new variants added at any time
#[derive(Clone, Debug, Error, PartialEq)]
#[non_exhaustive]
pub enum SemanticError {
    #[error("{0}")]
    Generic(String),

    // Declaration specifier errors
    #[error("cannot combine '{new}' specifier with previous '{existing}' type specifier")]
    InvalidSpecifier {
        existing: DeclarationSpecifier,
        new: DeclarationSpecifier,
    },
    #[error("'{0}' is not a qualifier and cannot be used for pointers")]
    NotAQualifier(DeclarationSpecifier),

    #[error("'{}' is too long for {}", vec!["long"; *.0].join(" "), env!("CARGO_PKG_NAME"))]
    TooLong(usize),

    #[error("conflicting storage classes '{0}' and '{1}'")]
    ConflictingStorageClass(StorageClass, StorageClass),

    #[error("conflicting types '{0}' and '{1}'")]
    ConflictingType(TypeKind, TypeKind),

    #[error("'{0}' cannot be signed or unsigned")]
    CannotBeSigned(TypeKind),

    #[error("types cannot be both signed and unsigned")]
    ConflictingSigned,

    #[error("only function-scoped variables can have an `auto` storage class")]
    AutoAtGlobalScope,

    #[error("cannot have empty program")]
    EmptyProgram,

    // Declarator errors
    #[error("expected an integer")]
    NonIntegralLength,

    #[error("arrays must have a positive length")]
    NegativeLength,

    #[error("function parameters always have a storage class of `auto`")]
    ParameterStorageClass(StorageClass),

    #[error("duplicate parameter name '{0}' in function declaration")]
    DuplicateParameter(InternedStr),

    #[error("functions cannot return '{0}'")]
    IllegalReturnType(TypeKind),

    // TODO: print params in the error message
    #[error("arrays cannot contain functions (got '{0}'). help: try storing array of pointer to function: (*{{}}[])(...)")]
    ArrayStoringFunction(TypeKind),

    #[error("void must be the first and only parameter if specified")]
    InvalidVoidParameter,

    #[error("functions taking `void` must not have variadic arguments")]
    VoidVarargs,

    #[error("functions taking variadic arguments must have at least one parameter first")]
    VarargsWithoutParam,

    #[error("overflow in enumeration constant")]
    EnumOverflow,

    #[error("variable has incomplete type 'void'")]
    VoidType,

    // expression errors
    #[error("use of undeclared identifier '{0}'")]
    UndeclaredVar(InternedStr),

    #[error("expected expression, got typedef")]
    TypedefInExpressionContext,

    #[error("type casts cannot have a storage class")]
    IllegalStorageClass(StorageClass),

    #[error("type casts cannot have a variable name")]
    IdInTypeName(InternedStr),

    #[error("expected integer, got '{0}'")]
    NonIntegralExpr(TypeKind),

    #[error("cannot implicitly convert '{0}' to '{1}'{}",
        if .1.is_pointer() {
            format!(". help: use an explicit cast: ({})", .1)
        } else {
            String::new()
        })
    ]
    InvalidCast(TypeKind, TypeKind),

    // String is the reason it couldn't be assigned
    #[error("cannot assign to {0}")]
    NotAssignable(String),

    #[error("invalid operators for '{0}' (expected either arithmetic types or pointer operation, got '{1} {0} {2}'")]
    InvalidAdd(BinaryOp, TypeKind, TypeKind),

    #[error("cannot perform pointer arithmetic when size of pointed type '{0}' is unknown")]
    PointerAddUnknownSize(TypeKind),

    #[error("called object of type '{0}' is not a function")]
    NotAFunction(TypeKind),

    #[error("too {} arguments to function call: expected {0}, have {1}", if .1 > .0 { "many" } else { "few" })]
    /// (actual, expected)
    WrongArgumentNumber(usize, usize),

    #[error("{0} has not yet been defined")]
    IncompleteDefinitionUsed(TypeKind),

    #[error("no member named '{0}' in '{1}'")]
    NotAMember(InternedStr, TypeKind),

    #[error("expected struct or union, got type '{0}'")]
    NotAStruct(TypeKind),

    #[error("cannot use '->' operator on type that is not a pointer")]
    NotAStructPointer(TypeKind),

    #[error("cannot dereference expression of non-pointer type '{0}'")]
    NotAPointer(TypeKind),

    #[error("cannot take address of {0}")]
    InvalidAddressOf(&'static str),

    #[error("cannot increment or decrement value of type '{0}'")]
    InvalidIncrement(TypeKind),

    #[error("cannot use unary plus on expression of non-arithmetic type '{0}'")]
    NotArithmetic(TypeKind),

    #[error("incompatible types in ternary expression: '{0}' cannot be converted to '{1}'")]
    IncompatibleTypes(TypeKind, TypeKind),

    // const fold errors
    #[error("{} overflow in expression", if *(.is_positive) { "positive" } else { "negative" })]
    ConstOverflow { is_positive: bool },

    #[error("cannot divide by zero")]
    DivideByZero,

    #[error("cannot shift {} by a negative amount", if *(.is_left) { "left" } else { "right" })]
    NegativeShift { is_left: bool },

    #[error("cannot shift {} by {maximum} or more bits for type '{ctype}' (got {current})",
        if *(.is_left) { "left" } else { "right" })]
    TooManyShiftBits {
        is_left: bool,
        maximum: u64,
        ctype: TypeKind,
        current: u64,
    },

    #[error("not a constant expression: {0}")]
    NotConstant(crate::hir::Expr),

    #[error("cannot dereference NULL pointer")]
    NullPointerDereference,

    #[error("invalid types for '{0}' (expected arithmetic types or compatible pointers, got {1} {0} {2}")]
    InvalidRelationalType(ComparisonToken, TypeKind, TypeKind),

    #[error("cannot cast pointer to float or vice versa")]
    FloatPointerCast(TypeKind),

    // TODO: this shouldn't be an error
    #[error("cannot cast to non-scalar type '{0}'")]
    NonScalarCast(TypeKind),

    #[error("cannot cast void to any type")]
    VoidCast,

    #[error("cannot cast structs to any type")]
    StructCast,

    // Control flow errors
    #[error("unreachable statement")]
    UnreachableStatement,

    // TODO: this error should happen way before codegen
    // #[cfg(feature = "codegen")]
    // #[error("redeclaration of label {0}")]
    // LabelRedeclaration(cranelift_codegen::ir::entities::Block),
    #[error("use of undeclared label {0}")]
    UndeclaredLabel(InternedStr),

    #[error("{}case outside of switch statement", if *(.is_default) { "default " } else { "" })]
    CaseOutsideSwitch { is_default: bool },

    #[error("cannot have multiple {}cases in a switch statement",
            if *(.is_default) { "default " } else { "" } )]
    DuplicateCase { is_default: bool },

    // Initializer errors
    #[error("initializers cannot be empty")]
    EmptyInitializer,

    #[error("scalar initializers for '{0}' may only have one element (initialized with {1})")]
    AggregateInitializingScalar(TypeKind, usize),

    #[error("too many initializers (declared with {0} elements, found {1})")]
    TooManyMembers(usize, usize),

    // Function definition errors
    #[error("illegal storage class {0} for function (only `static` and `extern` are allowed)")]
    InvalidFuncStorageClass(StorageClass),

    #[error("missing parameter name in function definition (parameter {0} of type '{1}')")]
    MissingParamName(usize, TypeKind),

    #[error("forward declaration of {0} is never completed (used in {1})")]
    ForwardDeclarationIncomplete(InternedStr, InternedStr),

    #[error("illegal signature for main function (expected 'int main(void)' or 'int main(int, char **)'")]
    IllegalMainSignature,

    // declaration errors
    #[error("redefinition of '{0}'")]
    Redefinition(InternedStr),

    #[error("redeclaration of '{0}' with different type or qualifiers (originally {}, now {})", .1.get(), .2.get())]
    IncompatibleRedeclaration(InternedStr, Symbol, Symbol),

    #[error("'{0}' can only appear on functions")]
    FuncQualifiersNotAllowed(FunctionQualifiers),

    #[error("switch expressions must have an integer type (got {0})")]
    NonIntegralSwitch(TypeKind),

    #[error("function '{0}' does not return a value")]
    MissingReturnValue(InternedStr),

    #[error("void function '{0}' should not return a value")]
    ReturnFromVoid(InternedStr),
}
