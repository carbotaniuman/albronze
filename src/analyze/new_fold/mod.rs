#[cfg(test)]
mod tests;

use crate::data::{ComparisonToken, LiteralValue};
use crate::hir::{BinaryOp, Expr, ExprType, TypeKind};

use crate::location::{Locatable, Location};
use dyn_safe::dyn_safe;
use thiserror::Error;

#[derive(Clone, Debug, Error, PartialEq)]
pub enum FoldError {
    #[error("{} overflow in expression", if *(.is_positive) { "positive" } else { "negative" })]
    ConstOverflow { is_positive: bool },

    #[error("cannot divide by zero")]
    DivideByZero,

    #[error("cannot shift {} by a negative amount", if *(.is_left) { "left" } else { "right" })]
    NegativeShift { is_left: bool },

    #[error("cannot shift {} by {maximum} or more bits for type '{ctype}' (got {current})",
        if *(.is_left) { "left" } else { "right" })]
    TooManyShiftBits {
        is_left: bool,
        maximum: u64,
        ctype: TypeKind,
        current: u64,
    },

    #[error("the given expression is invalid in this context")]
    Invalid,
}

/// Trait for a common interface for all constant expression evaluators.
#[dyn_safe(true)]
pub trait Folder {
    /// Folds the given `expr` into a single [`Expr`] with type
    /// [`ExprType::Literal`], or errors. This funtion returns the original
    /// expression in case of failure.
    fn const_fold(&self, expr: &Expr) -> Result<Expr, Locatable<FoldError>>;
}

impl<T: Folder> Folder for &T {
    fn const_fold(&self, expr: &Expr) -> Result<Expr, Locatable<FoldError>> {
        (*self).const_fold(expr)
    }
}

pub struct NoopFolder;

impl Folder for NoopFolder {
    fn const_fold(&self, expr: &Expr) -> Result<Expr, Locatable<FoldError>> {
        Err(expr.location.with(FoldError::Invalid))
    }
}

pub struct PreprocessorFolder;

impl Folder for PreprocessorFolder {
    fn const_fold(&self, expr: &Expr) -> Result<Expr, Locatable<FoldError>> {
        let location = expr.location;

        use LiteralValue::*;
        let folded = match &expr.expr {
            a @ ExprType::Literal(Int(_) | Char(_)) => a.clone(),
            ExprType::Literal(Float(_) | String(_)) => {
                return Err(expr.location.with(FoldError::Invalid))
            }
            ExprType::Negate(inner) => {
                map_literal(self, inner, location, |literal| match literal {
                    Int(i) => {
                        let (value, overflowed) = i.overflowing_neg();
                        if overflowed {
                            Err(FoldError::ConstOverflow { is_positive: i > 0 })
                        } else {
                            Ok(Int(value))
                        }
                    }
                    UnsignedInt(u) => {
                        // let signed = i64::try_from(big_number).map_err(|_| )?;
                        todo!()
                    }
                    Char(c) => Ok(Char(c.wrapping_neg())),
                    _ => unreachable!(),
                })?
            }
            ExprType::BitwiseNot(inner) => {
                map_literal(self, inner, location, |literal| match literal {
                    Int(i) => Ok(Int(!i)),
                    UnsignedInt(u) => Ok(UnsignedInt(!u)),
                    Char(c) => Ok(Char(!c)),
                    _ => unreachable!("bitwise not on invalid type"),
                })?
            }
            ExprType::Binary(op, left, right) => fold_binary(self, &left, &right, *op, location)?,
            ExprType::Ternary(condition, then, otherwise) => {
                let condition = self.const_fold(&condition)?;

                match condition.expr {
                    ExprType::Literal(Int(0) | UnsignedInt(0)) => self.const_fold(&otherwise)?.expr,
                    ExprType::Literal(Int(_) | UnsignedInt(_)) => self.const_fold(&then)?.expr,
                    _ => unreachable!(),
                }
            }
            ExprType::Comma(..) | ExprType::FuncCall(..) | ExprType::Stmt(..) => {
                return Err(expr.location.with(FoldError::Invalid))
            }
            ExprType::Cast(inner) if expr.ctype == TypeKind::Bool => bool_cast(self, inner)?,
            d => unreachable!("{:?} should have been caught by preprocessor", d),
        };

        Ok(Expr {
            expr: folded,
            ctype: expr.ctype.clone(),
            lval: expr.lval,
            location,
        })
    }
}

fn handle_bin_op<M>(left: &ExprType, right: &ExprType, mapper: M) -> Result<ExprType, FoldError>
where
    M: FnOnce(&LiteralValue, &LiteralValue) -> Result<LiteralValue, FoldError>,
{
    let left = match left {
        ExprType::Literal(literal) => literal,
        _ => unreachable!(),
    };
    let right = match right {
        ExprType::Literal(literal) => literal,
        _ => unreachable!(),
    };

    Ok(ExprType::Literal(mapper(left, right)?))
}

macro_rules! fold_int_bin_op {
    ($op: tt) => {
        |a: &LiteralValue, b: &LiteralValue| {
            use LiteralValue::*;
            match (a, b) {
                (Int(a), Int(b)) => Ok(Int(a $op b)),
                (UnsignedInt(a), UnsignedInt(b)) => Ok(UnsignedInt(a $op b)),
                (Char(a), Char(b)) => Ok(Char(a $op b)),
                (_, _) => todo!(),
            }
        }
    }
}

macro_rules! fold_compare_op {
    ($left: expr, $right: expr, $constructor: ident, $op: tt) => {{
            use LiteralValue::*;

            let left: &ExprType = $left;
            let right: &ExprType = $right;
            match (left, right) {
                (ExprType::Literal(a), ExprType::Literal(b)) => {
                    match (a, b) {
                        (Int(a), Int(b)) => ExprType::Literal(Int((a $op b) as i64)),
                        (UnsignedInt(a), UnsignedInt(b)) => ExprType::Literal(Int((a $op b) as i64)),
                        #[allow(clippy::float_cmp)]
                        (Float(a), Float(b)) => ExprType::Literal(Int((a $op b) as i64)),
                        (Char(a), Char(b)) => ExprType::Literal(Int((a $op b) as i64)),
                        (_, _) => todo!(),
                    }
                }
                _ => todo!(),
            }
        }}
    }

fn fold_binary(
    folder: impl Folder,
    left: &Expr,
    right: &Expr,
    op: BinaryOp,
    location: Location,
) -> Result<ExprType, Locatable<FoldError>> {
    let left = folder.const_fold(left)?;
    let right = folder.const_fold(right)?;

    use std::ops::*;
    use BinaryOp::*;
    use ComparisonToken::*;
    let expr_type = match op {
        Add => handle_bin_op(
            &left.expr,
            &right.expr,
            fold_scalar_bin_op(
                f64::add,
                i64::overflowing_add,
                u64::wrapping_add,
                u8::wrapping_add,
            ),
        ),
        Sub => handle_bin_op(
            &left.expr,
            &right.expr,
            fold_scalar_bin_op(
                f64::sub,
                i64::overflowing_sub,
                u64::wrapping_sub,
                u8::wrapping_sub,
            ),
        ),
        Mul => handle_bin_op(
            &left.expr,
            &right.expr,
            fold_scalar_bin_op(
                f64::mul,
                i64::overflowing_mul,
                u64::wrapping_mul,
                u8::wrapping_mul,
            ),
        ),
        Div => {
            if right.ctype.is_integral() && right.is_zero() {
                return Err(location.with(FoldError::DivideByZero));
            }
            handle_bin_op(
                &left.expr,
                &right.expr,
                fold_scalar_bin_op(
                    f64::div,
                    i64::overflowing_div,
                    u64::wrapping_div,
                    u8::wrapping_div,
                ),
            )
        }
        Mod => {
            if right.is_zero() {
                return Err(location.with(FoldError::DivideByZero));
            }

            use LiteralValue::*;
            handle_bin_op(
                &left.expr,
                &right.expr,
                |a: &LiteralValue, b: &LiteralValue| match (a, b) {
                    (Int(a), Int(b)) => {
                        let (value, overflowed) = a.overflowing_rem(*b);

                        if overflowed {
                            Err(FoldError::ConstOverflow {
                                is_positive: value.is_negative(),
                            })
                        } else {
                            Ok(Int(value))
                        }
                    }
                    (UnsignedInt(a), UnsignedInt(b)) => Ok(UnsignedInt(a.wrapping_rem(*b))),
                    (_, _) => unreachable!(),
                },
            )
        }
        Xor => handle_bin_op(&left.expr, &right.expr, fold_int_bin_op!(^)),
        BitwiseAnd => handle_bin_op(&left.expr, &right.expr, fold_int_bin_op!(&)),
        BitwiseOr => handle_bin_op(&left.expr, &right.expr, fold_int_bin_op!(|)),
        //     Shl => shift_left(left, right, parent_type, &location),
        //     Shr => shift_right(left, right, parent_type, &location),
        LogicalAnd => handle_bin_op(&left.expr, &right.expr, |left, right| {
            use LiteralValue::*;
            match (left, right) {
                (Int(1), Int(1)) => Ok(Int(1)),
                (Int(0), _) | (_, Int(0)) => Ok(Int(0)),
                _ => todo!(),
            }
        }),
        LogicalOr => handle_bin_op(&left.expr, &right.expr, |left, right| {
            use LiteralValue::*;
            match (left, right) {
                (Int(0), Int(0)) => Ok(Int(0)),
                (Int(1), _) | (_, Int(1)) => Ok(Int(1)),
                _ => todo!(),
            }
        }),
        Assign => Err(FoldError::Invalid),
        Compare(Less) => Ok(fold_compare_op!(&left.expr, &right.expr, Compare, <)),
        Compare(LessEqual) => Ok(fold_compare_op!(&left.expr, &right.expr, Compare, <=)),
        Compare(Greater) => Ok(fold_compare_op!(&left.expr, &right.expr, Compare, >)),
        Compare(GreaterEqual) => Ok(fold_compare_op!(&left.expr, &right.expr, Compare, >=)),
        Compare(EqualEqual) => Ok(fold_compare_op!(&left.expr, &right.expr, Compare, ==)),
        Compare(NotEqual) => Ok(fold_compare_op!(&left.expr, &right.expr, Compare, !=)),
        _ => todo!(),
    }
    .map_err(|e| location.with(e))?;

    Ok(expr_type)
}

#[inline]
fn fold_scalar_bin_op(
    simple: fn(f64, f64) -> f64,
    overflowing: fn(i64, i64) -> (i64, bool),
    wrapping: fn(u64, u64) -> u64,
    wrapping_byte: fn(u8, u8) -> u8,
) -> impl Fn(&LiteralValue, &LiteralValue) -> Result<LiteralValue, FoldError> {
    use LiteralValue::*;
    move |a: &LiteralValue, b: &LiteralValue| match (a, b) {
        (Int(a), Int(b)) => {
            // overflowing returns the wrapped value, so if we had a negative
            // value, it would be a positive overflow.
            let (value, overflowed) = overflowing(*a, *b);
            if overflowed {
                Err(FoldError::ConstOverflow {
                    is_positive: value.is_negative(),
                })
            } else {
                Ok(Int(value))
            }
        }
        (UnsignedInt(a), UnsignedInt(b)) => Ok(UnsignedInt(wrapping(*a, *b))),
        (Char(a), Char(b)) => Ok(Char(wrapping_byte(*a, *b))),
        (Float(a), Float(b)) => Ok(Float(simple(*a, *b))),
        // TODO: find a way to do this that allows `"hello" + 2 - 1`
        //(Str(s), Int(i)) | (Int(i), Str(s)) => {
        _ => todo!(),
    }
}

fn bool_cast(folder: impl Folder, expr: &Expr) -> Result<ExprType, Locatable<FoldError>> {
    let mut expr = folder.const_fold(expr)?;

    use LiteralValue::*;
    Ok(
        match expr.expr {
            ExprType::Literal(Int(i)) => ExprType::Literal(Int((i != 0) as i64)),
            ExprType::Literal(UnsignedInt(i)) => ExprType::Literal(UnsignedInt((i != 0) as u64)),
            _ => unreachable!(),
        }
    )
}

// fn cast(expr: Expr, ctype: &TypeKind) -> CompileResult<ExprType> {
//     let expr = expr.const_fold()?;
//     Ok(if let ExprType::Literal(ref token) = expr.expr {
//         if let Some(token) = const_cast(token, ctype) {
//             ExprType::Literal(token)
//         } else {
//             ExprType::Cast(Box::new(expr))
//         }
//     } else {
//         ExprType::Cast(Box::new(expr))
//     })
// }

// /// since we only have Int and Float for literals,
// /// all this does is make sure the folded value is in a valid range
// /// TODO: when we add suffix literals, that will have type information
// /// and we can use that to store the new type
// fn const_cast(token: &LiteralValue, ctype: &TypeKind) -> Option<LiteralValue> {
//     use LiteralValue::*;
//     let token = match (token, ctype) {
//         (Int(i), TypeKind::Bool) => Int((*i != 0).into()),
//         (Int(i), TypeKind::Char(_)) => Char(*i as u8),
//         (Int(i), TypeKind::Double) | (Int(i), TypeKind::Float) => Float(*i as f64),
//         (Int(i), ty) if ty.is_integral() && ty.is_signed() => Int(*i),
//         (Int(i), ty) if ty.is_integral() => UnsignedInt(*i as u64),

//         (UnsignedInt(u), TypeKind::Bool) => Int((*u != 0).into()),
//         (UnsignedInt(u), TypeKind::Char(_)) => Char(*u as u8),
//         (UnsignedInt(u), TypeKind::Double) | (UnsignedInt(u), TypeKind::Float) => Float(*u as f64),
//         (UnsignedInt(u), ty) if ty.is_integral() && ty.is_signed() => Int(*u as i64),
//         (UnsignedInt(u), ty) if ty.is_integral() => UnsignedInt(*u),

//         (Float(f), TypeKind::Bool) => Int((*f != 0.0) as i64),
//         (Float(f), TypeKind::Char(_)) => Char(*f as u8),
//         (Float(f), TypeKind::Double) | (Float(f), TypeKind::Float) => Float(*f),
//         (Float(f), ty) if ty.is_integral() && ty.is_signed() => Int(*f as i64),
//         (Float(f), ty) if ty.is_integral() => UnsignedInt(*f as u64),

//         (&Char(c), TypeKind::Bool) => Int((c != 0).into()),
//         (&Char(c), TypeKind::Double) | (&Char(c), TypeKind::Float) => Float(c.into()),
//         (&Char(c), ty) if ty.is_integral() && ty.is_signed() => Int(c.into()),
//         (&Char(c), ty) if ty.is_integral() => UnsignedInt(c.into()),

//         (Int(i), _) if ctype.is_pointer() && *i >= 0 => UnsignedInt(*i as u64),
//         (UnsignedInt(u), _) if ctype.is_pointer() => UnsignedInt(*u),
//         (&Char(c), _) if ctype.is_pointer() => UnsignedInt(c.into()),
//         _ => return None,
//     };
//     Some(token)
// }

fn map_literal<M>(
    folder: impl Folder,
    inner: &Expr,
    location: Location,
    mapper: M,
) -> Result<ExprType, Locatable<FoldError>>
where
    M: FnOnce(LiteralValue) -> Result<LiteralValue, FoldError>,
{
    let folded = folder.const_fold(inner)?;
    let literal = match folded.expr {
        ExprType::Literal(literal) => mapper(literal).map_err(|e| location.with(e))?,
        _ => unreachable!(),
    };

    Ok(ExprType::Literal(literal))
}
