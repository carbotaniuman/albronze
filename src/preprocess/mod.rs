pub mod error;
mod lexer;
mod manager;
mod preprocessor;
mod token;

pub use lexer::LexResult;
pub use manager::FileManager;
pub use preprocessor::Preprocessor;
pub use token::{pretty_print, TokenKind};
