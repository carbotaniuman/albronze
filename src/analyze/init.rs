//! 6.7.9 Initialization

use crate::analyze::error::SemanticError;
use crate::analyze::hir::{Expr, ExprType, Initializer, TypeKind};
use crate::location::Location;
use crate::parse::ast;

use super::PureAnalyzer;

impl PureAnalyzer {
    pub(super) fn parse_initializer(
        &mut self,
        init: ast::Initializer,
        ctype: &TypeKind,
        location: Location,
    ) -> Initializer {
        use ast::Initializer::{Aggregate, Scalar};
        // initializer_list
        let mut expr = match init {
            Aggregate(list) => return self.check_aggregate_overflow(list, ctype, location),
            Scalar(expr) => self.expr(*expr),
        };
        // The only time (that I know of) that an expression will initialize a non-scalar
        // is for character literals.
        let is_char_array = match ctype {
            TypeKind::Array(inner, _) => inner.is_char(),
            _ => false,
        };
        // See section 6.7.9 of the C11 standard:
        // The initializer for a scalar shall be a single expression, optionally enclosed in braces.
        // The initial value of the object is that of the expression (after conversion)
        if !is_char_array {
            expr = expr
                .rval()
                // if ctype is not a scalar, this will report an error, so we don't have to handle it specially
                .implicit_cast(ctype, &mut self.error_handler);
        }
        if !expr.lval && self.scope.is_global() && ctype.is_pointer() {
            expr = Expr {
                lval: false,
                location: expr.location,
                ctype: expr.ctype.clone(),
                expr: ExprType::StaticRef(Box::new(expr)),
            };
        }
        Initializer::Scalar(Box::new(expr))
    }

    fn check_aggregate_overflow(
        &mut self,
        list: Vec<ast::Initializer>,
        ctype: &TypeKind,
        location: Location,
    ) -> Initializer {
        let len = list.len();
        let mut iter = list.into_iter().peekable();
        let init = self.aggregate_initializer(&mut iter, ctype, location);
        let leftover = iter.count();
        if leftover > 0 {
            self.err(SemanticError::TooManyMembers(len - leftover, len), location);
        }
        init
    }

    // handle char[][3] = {{1,2,3}}, but also = {1,2,3} and {{1}, 2, 3}
    // NOTE: this does NOT consume {} except for sub-elements
    // see p17: "Each brace-enclosed initializer list has an associated current object"
    // For each subobject of the enclosing object (`type_at`), initialize it, possibly recursively.
    fn aggregate_initializer(
        &mut self,
        list: &mut std::iter::Peekable<impl Iterator<Item = ast::Initializer>>,
        elem_type: &TypeKind,
        location: Location,
    ) -> Initializer {
        use ast::Initializer::{Aggregate, Scalar};

        let mut elems = vec![];
        if list.peek().is_none() {
            self.err(SemanticError::EmptyInitializer, location);
            return Initializer::InitializerList(elems);
        }
        // char [][3] = {1};
        while let Some(elem) = list.peek() {
            let inner = elem_type.type_at(elems.len()).unwrap_or_else(|err| {
                // int a[1] = {1, 2};
                self.err(err, location);
                TypeKind::Error
            });
            // int a[][3] = {{1,2,3}}
            //               ^
            // initializer is aggregate, type errors will be caught later

            // If the initializer of a subaggregate or contained union begins with a left brace,
            // the initializers enclosed by that brace and its matching right brace initialize
            // the elements or members of the subaggregate or the contained union.
            let next = match elem {
                Aggregate(_) => match list.next() {
                    Some(Aggregate(inner_list)) => {
                        self.check_aggregate_overflow(inner_list, &inner, location)
                    }
                    _ => unreachable!(),
                },
                Scalar(_) => {
                    // int a[][3] = {1,2,3}
                    //               ^
                    // type is aggregate and initializer is scalar
                    // see if we can short circuit int[][3] -> int[3]
                    if inner != TypeKind::Error && !inner.is_scalar() {
                        // Note: this element is _not_ consumed
                        self.aggregate_initializer(list, &inner, location)
                    // type is scalar and initializer is scalar
                    // int a[][3] = {{1,2,3}}
                    } else {
                        let expr = match list.next() {
                            Some(Scalar(expr)) => self
                                .expr(*expr)
                                .rval()
                                .implicit_cast(&inner, &mut self.error_handler),
                            _ => unreachable!(),
                        };
                        Initializer::Scalar(Box::new(expr))
                    }
                }
            };
            elems.push(next);

            // Otherwise, only enough initializers from the list are taken
            // to account for the elements or members of the subaggregate
            // or the first member of the contained union;
            // any remaining initializers are left to initialize the next
            // element or member of the aggregate of which the current
            // subaggregate or contained union is a part.
            if elems.len() == elem_type.type_len() {
                break;
            }
        }
        Initializer::InitializerList(elems)
    }
}

impl TypeKind {
    /// Given a type, return the maximum number of initializers for that type
    fn type_len(&self) -> usize {
        use crate::analyze::hir::ArrayType;

        match self {
            ty if ty.is_scalar() => 1,
            TypeKind::Array(_, ArrayType::Fixed(size)) => *size as usize,
            TypeKind::Array(_, ArrayType::Unbounded) => 0,
            TypeKind::Struct(st) | TypeKind::Union(st) => st.members().len(),
            TypeKind::Function { .. } | TypeKind::Error => 1,
            _ => unimplemented!("type checking for {}", self),
        }
    }

    /// Given a type and an index,
    /// return the type expected at that index in the initializer.
    ///
    /// e.g. if `struct s { int i; float f; };` is in scope,
    /// `type_at(s, 0)` will be `int` and `type_at(s, 1)` will be `float`
    fn type_at(&self, index: usize) -> Result<TypeKind, SemanticError> {
        match self {
            ty if ty.is_scalar() => {
                if index == 0 {
                    Ok(ty.clone())
                } else {
                    Err(SemanticError::AggregateInitializingScalar(
                        ty.clone(),
                        index + 1,
                    ))
                }
            }
            TypeKind::Array(inner, _) => Ok((**inner).clone()),
            TypeKind::Struct(struct_type) => {
                let symbols = struct_type.members();
                symbols.get(index).map_or_else(
                    || Err(SemanticError::TooManyMembers(symbols.len(), index)),
                    |symbol| Ok(symbol.ctype.clone()),
                )
            }
            TypeKind::Union(struct_type) => {
                if index != 0 {
                    return Err("can only initialize first member of an enum".into());
                }
                let members = struct_type.members();
                Ok(members
                    .first()
                    .map(|m| m.ctype.clone())
                    .unwrap_or(TypeKind::Error))
            }
            TypeKind::Function { .. } | TypeKind::Error => Ok(TypeKind::Error),
            _ => unimplemented!("type checking for aggregate initializers of type {}", self),
        }
    }
}
