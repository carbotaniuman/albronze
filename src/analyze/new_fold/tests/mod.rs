use crate::analyze::{Folder, FoldError};
use crate::analyze::PureAnalyzer;
use crate::location::Locatable;
use crate::parse::Parser;
use crate::preprocess::{FileManager, Keyword, Preprocessor, TokenKind};
use crate::{get_str, Files, Source};
use arcstr::ArcStr;
use std::str::FromStr;

fn process(folder: impl Folder, contents: ArcStr) -> Result<crate::hir::ExprType, Locatable<FoldError>> {
    let mut files = Files::new();
    let id = files.add(
        "<generated>",
        Source {
            code: contents.clone(),
            path: concat!("<generated>").into(),
        },
    );

    let preprocessed = {
        let mut processor = Preprocessor::new(FileManager::new(), false);

        processor.preprocess_file(id, contents.clone());
        processor.pending_tokens
    };

    let parsed = {
        let mut parser = Parser::new(
            preprocessed
                .into_iter()
                .map(|t| {
                    if let TokenKind::Identifier(id) = t.data {
                        if let Ok(keyword) = Keyword::from_str(get_str!(id)) {
                            return t.location.with(TokenKind::Keyword(keyword));
                        }
                    }

                    t
                })
                .map(|t| Ok(t)),
            false,
        );

        let expr = parser.ternary_expr().expect("test should have only a conditional expression");
        if !parser.is_empty() {
            panic!("test should have only an expression");
        }
        expr
    };

    let analyzed = PureAnalyzer::new().expr(parsed);

    folder.const_fold(&analyzed).map(|a| a.expr)
}

macro_rules! define_test {
    ($name: ident, $folder : expr, $contents: literal) => {
        #[test]
        fn $name() {
            insta::assert_debug_snapshot!(process($folder, arcstr::literal!($contents)));
        }
    };
}

// Preprocessor Things
use crate::analyze::PreprocessorFolder;
define_test!(simple_literal, PreprocessorFolder, "5");
define_test!(addition, PreprocessorFolder, "5 + 5");
define_test!(subtraction, PreprocessorFolder, "5 - 5");
define_test!(multiplication, PreprocessorFolder, "5 * 5");
define_test!(division, PreprocessorFolder, "5 / 5");
define_test!(bitwise_not, PreprocessorFolder, "~5");

// Our AST construction is just busted, we'll fix this later
define_test!(logical_not_1, PreprocessorFolder, "!5");
define_test!(logical_not_2, PreprocessorFolder, "!0");
define_test!(ternary_1, PreprocessorFolder, "1 ? 2 : 3");
define_test!(ternary_2, PreprocessorFolder, "!1 ? 2 : 3");
