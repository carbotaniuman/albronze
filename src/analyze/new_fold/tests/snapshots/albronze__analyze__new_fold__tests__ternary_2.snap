---
source: src/analyze/new_fold/tests/mod.rs
expression: "process(PreprocessorFolder, arcstr::literal!(\"!1 ? 2 : 3\"))"
---
Ok(
    Literal(
        Int(
            3,
        ),
    ),
)
